### PR TITLE
nixos/nat: support nat reflection

### LINKED ISSUE #35132: nat.forwardPorts reachable only from outside
## Issue description

For example
```
          nat = {
            enable = true;

            externalInterface = "br0";
            externalIP = "55.55.55.55";

            internalInterfaces = [ "br1" ];
            internalIPs = [ "10.8.0.0/16" ];

            forwardPorts = [ { sourcePort = 44444; destination = "10.8.1.1:44444"; } ];
          };
```

Then ```55.55.55.55:44444``` can be connected from the external world, but not from the machine itself and not from other machines behind the NAT.

Perhaps it is a feature, but seems annoying

### PR DESCRIPTION
**Pull request by volth**
###### Motivation for this change

Attempt to fix #35132

endpoints defined in ```nat.forwardPorts``` can be connected from the Internet, but also
 - [x] they should be connectable from the host itself (using its public Internet address)
 - [x] they should be connectable from other hosts behind the NAT (as NAT provides Internet connectivity and the endpoint is connectable from the Internet, so it can be reached via any public proxy but not directly)

That requires a tricky technique called NAT reflection https://unix.stackexchange.com/questions/282086/how-does-nat-reflection-nat-loopback-work

### CODE REVIEW

**Comment by mogorman** [comment]
@volth would it fix this as well? https://github.com/NixOS/nixpkgs/issues/28721

**Comment by volth** [comment]
@mogorman I do not know :(

**Comment by volth** [comment]
My use case is simpler: one of the ```qemu``` virtual machines runs nix-cache, and I just want to access it using the same ```address:port``` from:
1. the Internet
2. the host machine of the VM
3. the VM itself
4. other VMs on the same host

so far, only ```1``` is possible (other options works via proxy), this fix adds ```2``` and it would be nice to discover how to do ```3``` and ```4```

File: nixos/modules/services/networking/nat.nix [hunk 1]
```diff
@@ -55,6 +55,16 @@ let
         -j DNAT --to-destination ${fwd.destination}
     '') cfg.forwardPorts}
 
+    # Allow to connect to ${cfg.externalIP}:${fwd.sourcePort} from the host itself
+    ${optionalString (cfg.externalIP != null) (
```
Comment (lschuermann) [inline nixos/modules/services/networking/nat.nix]: Maybe we could make another rule for all traffic coming from the loopback interface, so we got that covered even when `externalIP` isn't set? This solves the "connectable from the Host itself" for traffic in the usual ranges, e.g. `127.0.0.0/8`.

Comment (volth) [reply nixos/modules/services/networking/nat.nix]: I am not sure that loopback is involved here, source IP is ```eth0```'s, not ```127.0.0.1``` and it works without ```net.ipv4.conf.all.route_localnet``` 

Anyway, I agree that ```cfg.externalIP``` meaning is overloaded.

Comment (volth) [reply nixos/modules/services/networking/nat.nix]: Connectivity to```127.0.0.1:${fwd.sourcePort}``` is another issue, closer to @mogorman's

**Comment by volth** [comment]
ok, works for me.

**Comment by rnhmjoj** [comment]
An interesting option available in vyatta is to match by the interface name instead of providing the "loopback" address. I don't know if this is preferred in general but it's really useful if you have a dynamic address.

It could be implemented with a job that runs regularly and updates a set of addresses like `addr_eth1`, then you can use `-m set --match-set addr_pppoe1 dst` in the nat rule.

What do you think?

**Comment by volth** [comment]
@rnhmjoj that would raise questions like "how often should it _runs regularly_?" or may it be triggered by some event instead of regular checking.
Anyway, such a cron- or eventdriven- script is able to remove the the default values (if ```loopbackIP``` is set in ```configuration.nix```) from ```iptables``` and add its own. ```iptables``` is mutable unlike ```configuration.nix```

**Comment by rnhmjoj** [comment]
>"how often should it runs regularly?"
yeah, that's a problem. I'm trying to use a dhcpcd hook for now.

**Comment by rnhmjoj** [comment]
This is what I came up with. It should handle static addresses and DHCP automatically, beside providing the external address directly. I haven't managed to write a nixos test or tried it yet but I verified at least the addresses are getting added to the sets as expected.

```diff
diff --git a/nixos/modules/services/networking/nat.nix b/nixos/modules/services/networking/nat.nix
index da3827c35e6..48c3f1a3a92 100644
--- a/nixos/modules/services/networking/nat.nix
+++ b/nixos/modules/services/networking/nat.nix
@@ -28,6 +28,17 @@ let
     iptables -w -t nat -N nixos-nat-pre
     iptables -w -t nat -N nixos-nat-post
 
+    # Create a set of addresses for each interface statically
+    # configured. These will be used to match by interface name
+    # without IP address for NAT reflection. Other services such
+    # as dhcpcd may populate these dynamically.
+    ${concatMapStrings (iface: ''
+      ipset create ADDR_${iface.name} hash:net family inet || true
+      ${concatMapStrings (ip: ''
+        ipset add ADDR_${iface.name} ip.address || true
+      '') iface.ipv4.addresses}
+    '') (attrValues config.networking.interfaces)}
+
     # We can't match on incoming interface in POSTROUTING, so
     # mark packets coming from the external interfaces.
     ${concatMapStrings (iface: ''
@@ -54,29 +65,30 @@ let
         --dport ${builtins.toString fwd.sourcePort} \
         -j DNAT --to-destination ${fwd.destination}
 
-      ${concatMapStrings (loopbackip:
-        let
+      # Add user-specified addresses to the match set
+      ipset create ADDR_${cfg.externalInterface} hash:net family inet || true
+      ${concatMapStrings (addr: ''
+        ipset add ADDR_${cfg.externalInterface} ${addr} || true
+      '') fwd.loopbackIPs}
+
+      ${let
           m                = builtins.match "([0-9.]+):([0-9-]+)" fwd.destination;
           destinationIP    = if (m == null) then throw "bad ip:ports `${fwd.destination}'" else elemAt m 0;
           destinationPorts = if (m == null) then throw "bad ip:ports `${fwd.destination}'" else elemAt m 1;
+          addrSet          = "ADDR_${cfg.externalInterface}";
         in ''
-          # Allow connections to ${loopbackip}:${toString fwd.sourcePort} from the host itself
+          # Allow connections to ADDR_${cfg.externalInterface}:${toString fwd.sourcePort} from the host itself
           iptables -w -t nat -A OUTPUT \
-            -d ${loopbackip} -p ${fwd.proto} \
+            -m set --match-set ${addrSet} -p ${fwd.proto} \
             --dport ${builtins.toString fwd.sourcePort} \
             -j DNAT --to-destination ${fwd.destination}
 
-          # Allow connections to ${loopbackip}:${toString fwd.sourcePort} from other hosts behind NAT
+          # Allow connections to ADDR_${cfg.externalInterface}:${toString fwd.sourcePort} from other hosts behind NAT
           iptables -w -t nat -A nixos-nat-pre \
-            -d ${loopbackip} -p ${fwd.proto} \
+            -m set --match-set ${addrSet} -p ${fwd.proto} \
             --dport ${builtins.toString fwd.sourcePort} \
             -j DNAT --to-destination ${fwd.destination}
-
-          iptables -w -t nat -A nixos-nat-post \
-            -d ${destinationIP} -p ${fwd.proto} \
-            --dport ${destinationPorts} \
-            -j SNAT --to-source ${loopbackip}
-        '') fwd.loopbackIPs}
+        ''}
     '') cfg.forwardPorts}
 
     ${optionalString (cfg.dmzHost != null) ''
@@ -251,11 +263,18 @@ in
         extraStopCommands = flushNat;
       };
 
+      networking.dhcpcd.runHook = ''
+        if [ "$reason" = BOUND -o "$reason" = REBOOT ]; then
+            ${pkgs.ipset}/bin/ipset create ADDR_''${interface} hash:net family inet || true
+            ${pkgs.ipset}/bin/ipset add ADDR_''${interface} $new_ip_address  || true
+        fi
+      '';
+
       systemd.services = mkIf (!config.networking.firewall.enable) { nat = {
         description = "Network Address Translation";
         wantedBy = [ "network.target" ];
         after = [ "network-pre.target" "systemd-modules-load.service" ];
-        path = [ pkgs.iptables ];
+        path = with pkgs; [ iptables ipset ];
         unitConfig.ConditionCapability = "CAP_NET_ADMIN";
 
         serviceConfig = {

```

**Comment by volth** [comment]
```loopbackIPs``` are not always the addresses of the ```cfg.externaInterface```, they could be for example the IPs of Cloudflare if your service is behind Cloudflare.

**Pull request merged**